File: tests/formatters.test.ts
import { describe, test, expect } from 'bun:test';
import {
  formatSearchResults,
  formatAnswerResponse,
  formatResearchTask,
  formatError,
  formatSuccess,
} from '../src/formatters/markdown.js';

describe('formatSearchResults', () => {
  const mockSearchResponse = {
    requestId: 'test-123',
    costDollars: { total: 0.05 },
    results: [
      {
        title: 'Test Result',
        url: 'https://example.com',
        id: 'https://example.com',
        publishedDate: '2024-01-15',
        author: 'Test Author',
        score: 0.95,
        text: 'This is the content',
        highlights: ['highlight 1', 'highlight 2'],
        highlightScores: [0.9, 0.8],
        summary: 'This is a summary',
      },
    ],
  };

  test('formats search results in markdown', async () => {
    const output = await formatSearchResults(mockSearchResponse);
    expect(output).toContain('# Search Results');
    expect(output).toContain('Test Result');
    expect(output).toContain('https://example.com');
    expect(output).toContain('test-123');
    expect(output).toContain('$0.0500');
    expect(output).toContain('Test Author');
    expect(output).toContain('0.950');
    expect(output).toContain('This is the content');
    expect(output).toContain('highlight 1');
    expect(output).toContain('This is a summary');
  });

  test('returns JSON when json option is true', async () => {
    const output = await formatSearchResults(mockSearchResponse, {
      json: true,
    });
    expect(() => JSON.parse(output)).not.toThrow();
    const parsed = JSON.parse(output);
    expect(parsed.requestId).toBe('test-123');
  });

  test('handles empty results', async () => {
    const emptyResponse = { requestId: 'test-456', results: [] };
    const output = await formatSearchResults(emptyResponse);
    expect(output).toContain('No results found');
  });

  test('handles results without optional fields', async () => {
    const minimalResponse = {
      results: [
        {
          title: null,
          url: 'https://example.com',
          id: 'https://example.com',
        },
      ],
    };
    const output = await formatSearchResults(minimalResponse);
    expect(output).toContain('Untitled');
  });

  test('returns TOON when toon option is true', async () => {
    const output = await formatSearchResults(mockSearchResponse, {
      toon: true,
    });
    expect(typeof output).toBe('string');
    expect(output.length).toBeGreaterThan(0);
    expect(output.startsWith('#')).toBe(false); // not markdown
    expect(() => JSON.parse(output)).toThrow(); // not JSON
  });
});

describe('formatAnswerResponse', () => {
  const mockAnswerResponse = {
    requestId: 'answer-123',
    costDollars: { total: 0.02 },
    answer: 'This is the answer',
    citations: [
      { title: 'Source 1', url: 'https://source1.com' },
      { title: null, url: 'https://source2.com' },
    ],
  };

  test('formats answer in markdown', async () => {
    const output = await formatAnswerResponse(mockAnswerResponse);
    expect(output).toContain('# Answer');
    expect(output).toContain('This is the answer');
    expect(output).toContain('answer-123');
    expect(output).toContain('$0.0200');
    expect(output).toContain('Source 1');
    expect(output).toContain('Untitled');
  });

  test('formats object answers as JSON', async () => {
    const objectResponse = {
      answer: { key: 'value', number: 42 },
      citations: [],
    };
    const output = await formatAnswerResponse(objectResponse);
    expect(output).toContain('```json');
    expect(output).toContain('"key": "value"');
  });

  test('returns JSON when json option is true', async () => {
    const output = await formatAnswerResponse(mockAnswerResponse, {
      json: true,
    });
    const parsed = JSON.parse(output);
    expect(parsed.answer).toBe('This is the answer');
  });

  test('returns TOON when toon option is true', async () => {
    const output = await formatAnswerResponse(mockAnswerResponse, {
      toon: true,
    });
    expect(typeof output).toBe('string');
    expect(output.length).toBeGreaterThan(0);
    expect(output.startsWith('#')).toBe(false);
    expect(() => JSON.parse(output)).toThrow();
  });
});

describe('formatResearchTask', () => {
  const mockTask = {
    researchId: 'task-123',
    status: 'completed',
    instructions: 'Research something',
    costDollars: {
      total: 0.0059,
      numSearches: 1,
      numPages: 2,
      reasoningTokens: 100,
    },
    output: {
      parsed: { result: 'value' },
      content: 'Raw output content',
    },
    citations: [
      { url: 'https://example.com/1', title: 'Source One' },
      { url: 'https://example.com/2', title: null },
    ],
    events: [
      {
        createdAt: 1705315200000,
        eventType: 'research-start',
        message: 'Started research',
      },
    ],
  };

  test('formats research task in markdown', async () => {
    const output = await formatResearchTask(mockTask);
    expect(output).toContain('# Research Task');
    expect(output).toContain('task-123');
    expect(output).toContain('completed');
    expect(output).toContain('Research something');
    expect(output).toContain('"result": "value"');
    expect(output).toContain('Raw output content');
    expect(output).toContain('research-start');
  });

  test('formats citations as sources', async () => {
    const output = await formatResearchTask(mockTask);
    expect(output).toContain('## Sources');
    expect(output).toContain('[Source One](https://example.com/1)');
    expect(output).toContain('[Untitled](https://example.com/2)');
  });

  test('formats cost breakdown', async () => {
    const output = await formatResearchTask(mockTask);
    expect(output).toContain('$0.0059');
    expect(output).toContain('1 searches');
    expect(output).toContain('2 pages');
    expect(output).toContain('100 reasoning tokens');
  });

  test('handles tasks without output or events', async () => {
    const minimalTask = {
      researchId: 'task-456',
      status: 'pending',
    };
    const output = await formatResearchTask(minimalTask);
    expect(output).toContain('task-456');
    expect(output).toContain('pending');
    expect(output).not.toContain('## Output');
    expect(output).not.toContain('## Events');
  });

  test('returns JSON when json option is true', async () => {
    const output = await formatResearchTask(mockTask, { json: true });
    const parsed = JSON.parse(output);
    expect(parsed.researchId).toBe('task-123');
  });

  test('returns TOON when toon option is true', async () => {
    const output = await formatResearchTask(mockTask, { toon: true });
    expect(typeof output).toBe('string');
    expect(output.length).toBeGreaterThan(0);
    expect(output.startsWith('#')).toBe(false);
    expect(() => JSON.parse(output)).toThrow();
  });
});

describe('formatError', () => {
  test('formats Error instances', async () => {
    const error = new Error('Something went wrong');
    expect(await formatError(error)).toBe('Error: Something went wrong');
  });

  test('formats non-Error values', async () => {
    expect(await formatError('string error')).toBe('Error: string error');
    expect(await formatError(42)).toBe('Error: 42');
    expect(await formatError(null)).toBe('Error: null');
    expect(await formatError(undefined)).toBe('Error: undefined');
  });

  test('returns TOON when toon option is true', async () => {
    const output = await formatError(new Error('Something went wrong'), {
      toon: true,
    });
    expect(typeof output).toBe('string');
    expect(output.length).toBeGreaterThan(0);
    expect(output.startsWith('Error:')).toBe(false);
  });
});

describe('formatSuccess', () => {
  test('formats success messages', async () => {
    expect(await formatSuccess('Task completed')).toBe('[OK] Task completed');
    expect(await formatSuccess('Operation successful')).toBe(
      '[OK] Operation successful'
    );
  });

  test('returns TOON when toon option is true', async () => {
    const output = await formatSuccess('Task completed', { toon: true });
    expect(typeof output).toBe('string');
    expect(output.length).toBeGreaterThan(0);
    expect(output.startsWith('[OK]')).toBe(false);
  });
});

File: internal/formatters/formatters_test.go
package formatters

import (
	"encoding/json"
	"errors"
	"strings"
	"testing"

	"github.com/wesbragagt/exacli/internal/client"
)

// ---------------------------------------------------------------------------
// FormatSearchResults
// ---------------------------------------------------------------------------

func TestFormatSearchResults(t *testing.T) {
	resp := &client.SearchResponse{
		RequestID:   "req-123",
		CostDollars: client.CostDollars{Total: 0.0025},
		Results: []client.SearchResult{
			{
				Title:           "Example Page",
				URL:             "https://example.com",
				ID:              "id-1",
				PublishedDate:   "2024-01-15",
				Author:          "Alice",
				Score:           0.95,
				Text:            "Some content here.",
				Highlights:      []string{"highlighted text"},
				HighlightScores: []float64{0.87},
				Summary:         "A brief summary.",
			},
		},
	}

	t.Run("json mode", func(t *testing.T) {
		out := FormatSearchResults(resp, true, false)
		var parsed map[string]any
		if err := json.Unmarshal([]byte(out), &parsed); err != nil {
			t.Fatalf("JSON output is not valid JSON: %v", err)
		}
		if !strings.Contains(out, "req-123") {
			t.Error("JSON output missing requestId")
		}
		if !strings.Contains(out, "Example Page") {
			t.Error("JSON output missing result title")
		}
	})

	t.Run("toon mode", func(t *testing.T) {
		out := FormatSearchResults(resp, false, true)
		if !strings.Contains(out, "requestId: req-123") {
			t.Errorf("TOON output missing requestId line, got:\n%s", out)
		}
		if !strings.Contains(out, "results.0.url: https://example.com") {
			t.Errorf("TOON output missing results.0.url line, got:\n%s", out)
		}
	})

	t.Run("markdown mode", func(t *testing.T) {
		out := FormatSearchResults(resp, false, false)
		for _, want := range []string{
			"# Search Results",
			"Example Page",
			"https://example.com",
			"Request ID: req-123",
		} {
			if !strings.Contains(out, want) {
				t.Errorf("markdown output missing %q", want)
			}
		}
	})

	t.Run("empty results", func(t *testing.T) {
		empty := &client.SearchResponse{
			CostDollars: client.CostDollars{Total: 0},
		}
		out := FormatSearchResults(empty, false, false)
		if !strings.Contains(out, "# Search Results") {
			t.Error("empty results missing header")
		}
	})
}

// ---------------------------------------------------------------------------
// FormatAnswerResponse
// ---------------------------------------------------------------------------

func TestFormatAnswerResponse(t *testing.T) {
	resp := &client.AnswerResponse{
		RequestID:   "ans-456",
		CostDollars: client.CostDollars{Total: 0.005},
		Answer:      "The answer is 42.",
		Citations: []client.Citation{
			{Title: "Deep Thought", URL: "https://example.com/42"},
		},
	}

	t.Run("markdown with citations", func(t *testing.T) {
		out := FormatAnswerResponse(resp, false, false)
		for _, want := range []string{
			"# Answer",
			"The answer is 42.",
			"## Citations",
			"[Deep Thought](https://example.com/42)",
		} {
			if !strings.Contains(out, want) {
				t.Errorf("markdown output missing %q", want)
			}
		}
	})

	t.Run("json mode", func(t *testing.T) {
		out := FormatAnswerResponse(resp, true, false)
		var parsed map[string]any
		if err := json.Unmarshal([]byte(out), &parsed); err != nil {
			t.Fatalf("JSON output is not valid JSON: %v", err)
		}
		if !strings.Contains(out, "The answer is 42.") {
			t.Error("JSON output missing answer text")
		}
	})

	t.Run("no citations", func(t *testing.T) {
		noCitations := &client.AnswerResponse{
			RequestID:   "ans-789",
			CostDollars: client.CostDollars{Total: 0.001},
			Answer:      "No sources needed.",
		}
		out := FormatAnswerResponse(noCitations, false, false)
		if strings.Contains(out, "## Citations") {
			t.Error("output should not contain Citations section when there are none")
		}
	})
}

// ---------------------------------------------------------------------------
// FormatResearchTask
// ---------------------------------------------------------------------------

func TestFormatResearchTask(t *testing.T) {
	task := &client.ResearchTask{
		ResearchID:   "res-001",
		Status:       "completed",
		Instructions: "Analyze market trends",
		CostDollars: client.ResearchCostDollars{
			Total:           0.15,
			NumSearches:     5,
			NumPages:        20,
			ReasoningTokens: 1500,
		},
		Output: &client.ResearchOutput{
			Content: "Market analysis complete.",
		},
		Citations: []client.Citation{
			{Title: "Market Report", URL: "https://example.com/report"},
		},
		Events: []client.ResearchEvent{
			{
				CreatedAt: 1700000000000, // 2023-11-14T22:13:20Z
				EventType: "search",
				Message:   "Searching for trends",
			},
		},
	}

	t.Run("markdown full", func(t *testing.T) {
		out := FormatResearchTask(task, false, false)
		for _, want := range []string{
			"# Research Task",
			"res-001",
			"completed",
			"## Output",
			"Market analysis complete.",
			"## Events",
			"search",
			"Searching for trends",
		} {
			if !strings.Contains(out, want) {
				t.Errorf("markdown output missing %q, got:\n%s", want, out)
			}
		}
	})

	t.Run("with output content", func(t *testing.T) {
		out := FormatResearchTask(task, false, false)
		if !strings.Contains(out, "## Output") {
			t.Error("missing Output section")
		}
		if !strings.Contains(out, "Market analysis complete.") {
			t.Error("missing output content")
		}
	})

	t.Run("nil output", func(t *testing.T) {
		noOutput := &client.ResearchTask{
			ResearchID: "res-002",
			Status:     "processing",
			CostDollars: client.ResearchCostDollars{
				Total: 0.01,
			},
		}
		out := FormatResearchTask(noOutput, false, false)
		if strings.Contains(out, "## Output") {
			t.Error("should not contain Output section when output is nil")
		}
	})

	t.Run("events with timestamp", func(t *testing.T) {
		out := FormatResearchTask(task, false, false)
		// Timestamp should be formatted as RFC3339 in UTC
		if !strings.Contains(out, "2023-11-14T22:13:20Z") {
			t.Errorf("missing formatted timestamp in events, got:\n%s", out)
		}
	})
}

// ---------------------------------------------------------------------------
// FormatError
// ---------------------------------------------------------------------------

func TestFormatError(t *testing.T) {
	err := errors.New("something went wrong")

	cases := []struct {
		name   string
		asToon bool
		want   string
	}{
		{"plain", false, "Error: something went wrong\n"},
		{"toon", true, "error: something went wrong\n"},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := FormatError(err, tc.asToon)
			if got != tc.want {
				t.Errorf("FormatError(%v, %t) = %q, want %q", err, tc.asToon, got, tc.want)
			}
		})
	}
}

// ---------------------------------------------------------------------------
// FormatSuccess
// ---------------------------------------------------------------------------

func TestFormatSuccess(t *testing.T) {
	cases := []struct {
		name   string
		msg    string
		asToon bool
		want   string
	}{
		{"plain", "Task complete", false, "Task complete\n"},
		{"toon", "Task complete", true, "success: Task complete\n"},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := FormatSuccess(tc.msg, tc.asToon)
			if got != tc.want {
				t.Errorf("FormatSuccess(%q, %t) = %q, want %q", tc.msg, tc.asToon, got, tc.want)
			}
		})
	}
}

// ---------------------------------------------------------------------------
// FormatCodeContextResult
// ---------------------------------------------------------------------------

func TestFormatCodeContextResult(t *testing.T) {
	resp := &client.CodeContextResponse{
		RequestID: "ctx-001",
		Query:     "how to use cobra",
		Response:  "func main() { fmt.Println(\"hello\") }",
	}

	t.Run("markdown mode", func(t *testing.T) {
		out := FormatCodeContextResult(resp, false, false)
		for _, want := range []string{
			"# Code Context",
			"Request ID: ctx-001",
			"Query: how to use cobra",
			"func main()",
		} {
			if !strings.Contains(out, want) {
				t.Errorf("markdown output missing %q, got:\n%s", want, out)
			}
		}
	})

	t.Run("json mode", func(t *testing.T) {
		out := FormatCodeContextResult(resp, true, false)
		var parsed map[string]any
		if err := json.Unmarshal([]byte(out), &parsed); err != nil {
			t.Fatalf("JSON output is not valid JSON: %v", err)
		}
		if !strings.Contains(out, "ctx-001") {
			t.Error("JSON output missing requestId")
		}
	})

	t.Run("toon mode", func(t *testing.T) {
		out := FormatCodeContextResult(resp, false, true)
		if !strings.Contains(out, "requestId: ctx-001") {
			t.Errorf("TOON output missing requestId line, got:\n%s", out)
		}
	})

	t.Run("empty response does not panic", func(t *testing.T) {
		empty := &client.CodeContextResponse{}
		out := FormatCodeContextResult(empty, false, false)
		if !strings.Contains(out, "# Code Context") {
			t.Error("empty response missing header")
		}
	})
}
